src/main/java/com/google/api/codegen/transformer/ruby/RubyGapicSurfaceTestTransformer.java

```diff
@@ -0,0 +1,165 @@
+
+  public RubyGapicSurfaceTestTransformer(GapicCodePathMapper rubyPathMapper) {
+    this.pathMapper = rubyPathMapper;
+  }
+
+  @Override
+  public List<String> getTemplateFileNames() {
+    return ImmutableList.of(SMOKE_TEST_TEMPLATE_FILE);
+  }
+
+  @Override
+  public List<ViewModel> transform(Model model, ApiConfig apiConfig) {
+    List<ViewModel> views = new ArrayList<>();
+    for (Interface service : new InterfaceView().getElementIterable(model)) {
+      SurfaceTransformerContext context = createContext(service, apiConfig);
+      if (context.getInterfaceConfig().getSmokeTestConfig() != null) {
+        views.add(createSmokeTestClassView(context));
+      }
+    }
+
+    return views;
+  }
+
+  ///////////////////////////////////// Smoke Test ///////////////////////////////////////
+
+  private SmokeTestClassView createSmokeTestClassView(SurfaceTransformerContext context) {
+    Interface service = context.getInterface();
+    String outputPath = pathMapper.getOutputPath(service, context.getApiConfig());
+    SurfaceNamer namer = context.getNamer();
+    String name = namer.getSmokeTestClassName(service);
+
+    Method method = context.getInterfaceConfig().getSmokeTestConfig().getMethod();
+    FlatteningConfig flatteningGroup =
+        testCaseTransformer.getSmokeTestFlatteningGroup(
+            context.getMethodConfig(method), context.getInterfaceConfig().getSmokeTestConfig());
+    MethodTransformerContext flattenedMethodContext =
+        context.asFlattenedMethodContext(method, flatteningGroup);
+
+    SmokeTestClassView.Builder testClass = SmokeTestClassView.newBuilder();
+    // TODO: we need to remove testCaseView after we switch to use apiMethodView for smoke test
+    // testCaseView not in use by Ruby for smoke test.
+    TestCaseView testCaseView = testCaseTransformer.createSmokeTestCaseView(flattenedMethodContext);
+    OptionalArrayMethodView apiMethodView =
+        createSmokeTestCaseApiMethodView(flattenedMethodContext);
+
+    testClass.apiSettingsClassName(namer.getApiSettingsClassName(service));
+    testClass.apiClassName(namer.getApiWrapperClassName(service));
+    testClass.name(name);
+    testClass.outputPath(namer.getSourceFilePath(outputPath, name));
+    testClass.templateFileName(SMOKE_TEST_TEMPLATE_FILE);
+    testClass.apiMethod(apiMethodView);
+    testClass.method(testCaseView);
+    testClass.requireProjectId(
+        testCaseTransformer.requireProjectId(apiMethodView.initCode(), context.getNamer()));
+
+    FileHeaderView fileHeader = fileHeaderTransformer.generateFileHeader(context);
+    testClass.fileHeader(fileHeader);
+
+    return testClass.build();
+  }
+
+  private OptionalArrayMethodView createSmokeTestCaseApiMethodView(
+      MethodTransformerContext context) {
+    OptionalArrayMethodView initApiMethodView =
+        new DynamicLangApiMethodTransformer().generateMethod(context);
+
+    OptionalArrayMethodView.Builder apiMethodView = OptionalArrayMethodView.newBuilder();
```

Comment:
Add a `toBuilder()` method to `OptionalArrayMethodView` then you can get rid of the following lines.